File: src/features/notes/lib/markdownParser.ts
import { load } from 'js-yaml'
import type { HeadingItem, NoteMetadata, WikilinkItem } from '../types'

export function parseMarkdownFile(
  path: string,
  content: string,
  modifiedAt = Date.now(),
  size = 0,
): NoteMetadata {
  const lines = content.split(/\r?\n/)
  let frontmatter: Record<string, unknown> = {}
  let bodyLines = lines
  let bodyLineOffset = 0

  // 1. Extract YAML frontmatter if present
  if (lines.length > 0 && lines[0].trim() === '---') {
    const endIdx = lines.slice(1).findIndex((l) => l.trim() === '---')
    if (endIdx !== -1) {
      const yamlContent = lines.slice(1, endIdx + 1).join('\n')
      try {
        const parsed = load(yamlContent)
        if (parsed && typeof parsed === 'object') {
          frontmatter = parsed as Record<string, unknown>
        }
      } catch {
        // invalid yaml frontmatter, ignore
      }
      bodyLineOffset = endIdx + 2
      bodyLines = lines.slice(bodyLineOffset)
    }
  }

  // Determine Title: from frontmatter title, or first # H1, or filename
  const fileName = path.split('/').pop()?.replace(/\.md$/i, '') || 'Not'
  let title = typeof frontmatter.title === 'string' ? frontmatter.title : fileName

  const headings: HeadingItem[] = []
  const outgoingLinks: WikilinkItem[] = []
  const tagsSet = new Set<string>()

  // Extract tags from frontmatter
  if (Array.isArray(frontmatter.tags)) {
    for (const t of frontmatter.tags) {
      if (typeof t === 'string') {
        tagsSet.add(t.replace(/^#/, '').toLowerCase().trim())
      }
    }
  } else if (typeof frontmatter.tags === 'string') {
    for (const t of frontmatter.tags.split(',')) {
      const clean = t.replace(/^#/, '').toLowerCase().trim()
      if (clean) tagsSet.add(clean)
    }
  }

  let inCodeBlock = false

  // Regex for wikilinks: [[Target]] or [[Target|Alias]] or [[Target#Anchor]] or [[Target#Anchor|Alias]]
  const wikilinkRegex = /\[\[([^\]|#]+)(?:#([^\]|]+))?(?:\|([^\]]+))?\]\]/g

  // Regex for inline #tags: starts with # followed by alphanumeric/turkish chars/underscore/hyphen
  // Lookbehind or boundary to prevent matching URLs (http://...#hash) or hex colors
  const tagRegex = /(?:^|\s)(#[a-zA-Z0-9_\u00C0-\u017F\u0180-\u024F\u0400-\u04FF-]+)/g

  for (let i = 0; i < bodyLines.length; i++) {
    const line = bodyLines[i]
    const actualLineNumber = bodyLineOffset + i + 1

    if (line.trim().startsWith('```')) {
      inCodeBlock = !inCodeBlock
      continue
    }

    if (inCodeBlock) continue

    // Check headings (# H1 ... ###### H6)
    const headingMatch = line.match(/^(#{1,6})\s+(.+)$/)
    if (headingMatch) {
      const level = headingMatch[1].length
      const text = headingMatch[2].trim()
      headings.push({
        level,
        text,
        line: actualLineNumber,
      })
      if (level === 1 && !frontmatter.title && title === fileName) {
        title = text
      }
    }

    // Check wikilinks
    let linkMatch: RegExpExecArray | null
    wikilinkRegex.lastIndex = 0
    while ((linkMatch = wikilinkRegex.exec(line)) !== null) {
      const raw = linkMatch[0]
      const targetTitle = linkMatch[1]?.trim() || ''
      const targetAnchor = linkMatch[2]?.trim()
      const alias = linkMatch[3]?.trim()

      if (targetTitle) {
        // Get clean snippet around the link
        const contextSnippet = line.length > 180 ? line.slice(0, 180) + '...' : line

        outgoingLinks.push({
          raw,
          targetTitle,
          targetAnchor,
          alias,
          line: actualLineNumber,
          contextSnippet: contextSnippet.trim(),
        })
      }
    }

    // Check inline tags
    let tagMatch: RegExpExecArray | null
    tagRegex.lastIndex = 0
    while ((tagMatch = tagRegex.exec(line)) !== null) {
      const rawTag = tagMatch[1]
      // ensure it's not just # (which is a heading) or numeric only
      const cleanTag = rawTag.substring(1).trim().toLowerCase()
      if (cleanTag && !/^\d+$/.test(cleanTag)) {
        tagsSet.add(cleanTag)
      }
    }
  }

  return {
    path,
    title,
    frontmatter,
    headings,
    tags: Array.from(tagsSet),
    outgoingLinks,
    modifiedAt,
    size,
  }
}

export function normalizeNoteTitle(title: string): string {
  return title.trim().toLowerCase().replace(/\.md$/i, '')
}

File: mobil/app/src/main/java/com/kapanis/mobil/data/vault/MarkdownParser.kt
package com.kapanis.mobil.data.vault

import java.util.Locale

object MarkdownParser {

    private val wikilinkRegex = Regex("""\[\[([^\]|#]+)(?:#([^\]|]+))?(?:\|([^\]]+))?\]\]""")
    private val tagRegex = Regex("""(?:^|\s)(#[a-zA-Z0-9_\u00C0-\u017F\u0180-\u024F\u0400-\u04FF-]+)""")
    private val headingRegex = Regex("""^(#{1,6})\s+(.+)$""")

    fun normalizeTitle(title: String): String {
        return title.trim()
            .lowercase(Locale.getDefault())
            .removeSuffix(".md")
    }

    fun parseMarkdownFile(
        path: String,
        content: String,
        modifiedAt: Long = System.currentTimeMillis(),
        size: Long = content.length.toLong(),
        isPinned: Boolean = false
    ): VaultNote {
        val lines = content.lines()
        val frontmatter = mutableMapOf<String, Any>()
        var bodyLines = lines
        var bodyLineOffset = 0

        // 1. Extract YAML frontmatter if present
        if (lines.isNotEmpty() && lines[0].trim() == "---") {
            val endIdx = lines.drop(1).indexOfFirst { it.trim() == "---" }
            if (endIdx != -1) {
                val yamlLines = lines.subList(1, endIdx + 1)
                for (yLine in yamlLines) {
                    val colonIdx = yLine.indexOf(':')
                    if (colonIdx != -1) {
                        val key = yLine.substring(0, colonIdx).trim()
                        val value = yLine.substring(colonIdx + 1).trim().removeSurrounding("\"").removeSurrounding("'")
                        frontmatter[key] = value
                    }
                }
                bodyLineOffset = endIdx + 2
                bodyLines = if (bodyLineOffset < lines.size) lines.subList(bodyLineOffset, lines.size) else emptyList()
            }
        }

        // Determine filename & default title
        val fileName = path.substringAfterLast('/').removeSuffix(".md").ifEmpty { "Not" }
        var title = (frontmatter["title"] as? String)?.takeIf { it.isNotBlank() } ?: fileName

        val headings = mutableListOf<HeadingItem>()
        val outgoingLinks = mutableListOf<WikilinkItem>()
        val tagsSet = mutableSetOf<String>()

        // Check frontmatter tags
        val fmTags = frontmatter["tags"]
        if (fmTags is String) {
            fmTags.split(',', ';', ' ').forEach { t ->
                val clean = t.trim().removePrefix("#").lowercase(Locale.getDefault())
                if (clean.isNotEmpty()) tagsSet.add(clean)
            }
        }

        var inCodeBlock = false

        for ((i, line) in bodyLines.withIndex()) {
            val actualLineNumber = bodyLineOffset + i + 1
            val trimmed = line.trim()

            if (trimmed.startsWith("```")) {
                inCodeBlock = !inCodeBlock
                continue
            }

            if (inCodeBlock) continue

            // 1. Headings (# H1 ... ###### H6)
            val headingMatch = headingRegex.find(line)
            if (headingMatch != null) {
                val level = headingMatch.groupValues[1].length
                val text = headingMatch.groupValues[2].trim()
                headings.add(
                    HeadingItem(
                        level = level,
                        text = text,
                        line = actualLineNumber
                    )
                )
                if (level == 1 && frontmatter["title"] == null && title == fileName) {
                    title = text
                }
            }

            // 2. Wikilinks [[Target]]
            wikilinkRegex.findAll(line).forEach { match ->
                val raw = match.value
                val targetTitle = match.groupValues[1].trim()
                val targetAnchor = match.groupValues.getOrNull(2)?.trim()?.ifEmpty { null }
                val alias = match.groupValues.getOrNull(3)?.trim()?.ifEmpty { null }

                if (targetTitle.isNotEmpty()) {
                    val snippet = if (line.length > 150) line.take(150) + "..." else line
                    outgoingLinks.add(
                        WikilinkItem(
                            raw = raw,
                            targetTitle = targetTitle,
                            targetAnchor = targetAnchor,
                            alias = alias,
                            line = actualLineNumber,
                            contextSnippet = snippet.trim()
                        )
                    )
                }
            }

            // 3. Inline Tags #tag
            tagRegex.findAll(line).forEach { match ->
                val rawTag = match.groupValues[1]
                val cleanTag = rawTag.removePrefix("#").trim().lowercase(Locale.getDefault())
                if (cleanTag.isNotEmpty() && !cleanTag.all { it.isDigit() }) {
                    tagsSet.add(cleanTag)
                }
            }
        }

        return VaultNote(
            path = path,
            name = path.substringAfterLast('/'),
            title = title,
            content = content,
            frontmatter = frontmatter,
            headings = headings,
            tags = tagsSet.toList().sorted(),
            outgoingLinks = outgoingLinks,
            backlinks = emptyList(),
            modifiedAt = modifiedAt,
            size = size,
            isPinned = isPinned
        )
    }

    fun buildVaultIndex(notes: List<VaultNote>): VaultIndex {
        val filesMap = notes.associateBy { it.path }
        val titleToPathMap = mutableMapOf<String, String>()
        val outgoingLinksMap = mutableMapOf<String, List<WikilinkItem>>()
        val backlinksMap = mutableMapOf<String, MutableList<BacklinkItem>>()
        val tagsMap = mutableMapOf<String, MutableSet<String>>()

        for (note in notes) {
            val normTitle = normalizeTitle(note.title)
            val normFileName = normalizeTitle(note.name)
            titleToPathMap[normTitle] = note.path
            titleToPathMap[normFileName] = note.path

            outgoingLinksMap[note.path] = note.outgoingLinks

            for (tag in note.tags) {
                tagsMap.getOrPut(tag) { mutableSetOf() }.add(note.path)
            }
        }

        // Build backlinks
        for (note in notes) {
            for (link in note.outgoingLinks) {
                val targetNorm = normalizeTitle(link.targetTitle)
                val targetPath = titleToPathMap[targetNorm]
                if (targetPath != null && targetPath != note.path) {
                    val backlink = BacklinkItem(
                        sourcePath = note.path,
                        sourceTitle = note.title,
                        line = link.line,
                        contextSnippet = link.contextSnippet,
                        alias = link.alias
                    )
                    backlinksMap.getOrPut(targetPath) { mutableListOf() }.add(backlink)
                }
            }
        }

        // Attach backlinks to notes
        val enrichedFiles = filesMap.mapValues { (path, note) ->
            note.copy(backlinks = backlinksMap[path] ?: emptyList())
        }

        return VaultIndex(
            files = enrichedFiles,
            titleToPath = titleToPathMap,
            outgoingLinks = outgoingLinksMap,
            backlinks = backlinksMap,
            tags = tagsMap
        )
    }
}
